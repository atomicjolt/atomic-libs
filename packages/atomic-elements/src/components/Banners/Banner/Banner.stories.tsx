import { Meta, StoryObj } from "@storybook/react";
import { Banner } from ".";
import { MaterialIcon } from "../../Icons/MaterialIcon";
import { getCssProps } from "@sb/cssprops";
import { RenderPropsArgTypes } from "@sb/helpers";

export default {
  title: "Banners/Banner",
  component: Banner,
  parameters: {
    cssprops: getCssProps("Banner"),
  },
  argTypes: {
    ...RenderPropsArgTypes,
    variant: {
      control: "select",
      options: ["info", "error", "success", "warning"],
    },
  },
} as Meta<typeof Banner>;

type Story = StoryObj<typeof Banner>;

export const Primary: Story = {
  args: {
    variant: "info",
    children: <Banner.Content $ml="2">This is a banner</Banner.Content>,
  },
};

export const WithIcon: Story = {
  args: {
    ...Primary.args,
    children: [
      <MaterialIcon key="icon" icon="info" />,
      <Banner.Content key="content">This is a banner</Banner.Content>,
    ],
  },
};

export const WithIconButton: Story = {
  args: {
    ...Primary.args,
    children: [
      <MaterialIcon icon="info" key="icon" />,
      <Banner.Content key="content">This is a banner</Banner.Content>,
      <Banner.IconButton key="action" icon="close" />,
    ],
  },
};

export const WithButton: Story = {
  args: {
    ...Primary.args,
    children: [
      <MaterialIcon icon="info" />,
      <Banner.Content>This is a banner</Banner.Content>,
      <Banner.Button>Click Me</Banner.Button>,
    ],
  },
};
