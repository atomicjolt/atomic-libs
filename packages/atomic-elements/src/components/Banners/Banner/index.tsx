export { Banner } from "./Banner.component";
export type { BannerProps, BannerVariants } from "./Banner.component";
