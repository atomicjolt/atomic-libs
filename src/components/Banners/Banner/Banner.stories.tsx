import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import { Banner } from ".";
import MaterialIcon from "../../Icons/MaterialIcon";
import { getCssProps } from "@sb/cssprops";

export default {
  title: "Banners/Banner",
  component: Banner,
  parameters: {
    cssprops: getCssProps("Banner"),
  },
  argTypes: {
    variant: {
      control: "select",
      options: ["info", "error", "success", "warning"],
    },
  },
} as Meta<typeof Banner>;

type Story = StoryObj<typeof Banner>;

export const Primary: Story = {
  args: {
    children: "Content of the banner goes here",
  },
};

export const WithIcon: Story = {
  args: {
    children: [
      <MaterialIcon icon="info" />,
      <Banner.Content>This is a banner</Banner.Content>,
    ],
  },
};

export const WithIconButton: Story = {
  args: {
    children: [
      <MaterialIcon icon="info" />,
      <Banner.Content>This is a banner</Banner.Content>,
      <Banner.IconButton icon="close" />,
    ],
  },
};

export const WithButton: Story = {
  args: {
    children: [
      <MaterialIcon icon="info" />,
      <Banner.Content>This is a banner</Banner.Content>,
      <Banner.Button>Click Me</Banner.Button>,
    ],
  },
};
