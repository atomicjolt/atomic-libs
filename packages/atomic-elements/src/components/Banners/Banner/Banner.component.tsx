import React, { forwardRef, RefAttributes } from "react";
import { useRenderProps } from "@hooks/useRenderProps";
import {
  ElementWrapper,
  RenderBaseProps,
  SuggestStrings,
} from "../../../types";
import { BannerContent, StyledBanner } from "../Banners.styles";
import { IconButton, IconButtonProps } from "../../Buttons/IconButton";
import { Button, ButtonProps } from "../../Buttons/Button";

export type BannerVariants = SuggestStrings<
  "info" | "error" | "success" | "warning"
>;

export interface BannerProps
  extends RenderBaseProps<never>,
    ElementWrapper<HTMLDivElement> {
  readonly variant?: BannerVariants;
}

function Banner(props: BannerProps, ref: React.Ref<HTMLDivElement>) {
  const { variant = "info", children, className, style, ...rest } = props;

  const renderProps = useRenderProps({
    componentClassName: "aje-banner",
    variant,
    children,
    className,
    style,
  });

  return <StyledBanner ref={ref} {...renderProps} {...rest} />;
}

const BannerIconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  (props, ref) => {
    const { variant = "banner", ...rest } = props;
    return <IconButton ref={ref} variant={variant} {...rest} />;
  }
);

const BannerButton = forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const { variant = "inverted", ...rest } = props;
    return <Button ref={ref} variant={variant} {...rest} />;
  }
);

interface ForwardedBanner {
  (props: BannerProps & RefAttributes<HTMLDivElement>): JSX.Element;
  displayName: string;
  Content: typeof BannerContent;
  IconButton: typeof BannerIconButton;
  Button: typeof BannerButton;
}

/** A view-spanning Banner. The most Basic kind of banner,
 * with no associated actions / buttons. Useful building block when constructing
 * your own banners.
 */
const _Banner = forwardRef(Banner) as unknown as ForwardedBanner;
_Banner.displayName = "Banner";

/** Wrapper around the textual content of a Banner */
_Banner.Content = BannerContent;
BannerContent.displayName = "Banner.Content";

/** IconButton for Banners */
_Banner.IconButton = BannerIconButton;
BannerIconButton.displayName = "Banner.IconButton";

/** Button for Banners */
_Banner.Button = BannerButton;
BannerButton.displayName = "Banner.Button";

export { _Banner as Banner };
