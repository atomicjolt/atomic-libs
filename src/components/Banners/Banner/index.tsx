import React from "react";
import cn from "classnames";
import { HasChildren, HasClassName, SuggestStrings } from "../../../types";
import { BannerMain, StyledBanner } from "../Banners.styles";

export type BannerVariants = SuggestStrings<
  "info" | "error" | "success" | "warning"
>;

export interface BannerProps extends HasChildren, HasClassName {
  readonly variant?: BannerVariants;
  /** Render Something before the text content of the banner */
  readonly beforeContent?: React.ReactNode;
  /** Render Something after the text content of the banner */
  readonly afterContent?: React.ReactNode;
}

/** A view-spanning Banner. The most Basic kind of banner,
 * with no associated actions / buttons. Useful building block when constructing
 * your own banners.
 */
export default function Banner(props: BannerProps) {
  const {
    variant = "info",
    children,
    className,
    beforeContent,
    afterContent,
  } = props;

  return (
    <StyledBanner className={cn(`aje-banner--${variant}`, className)}>
      {beforeContent}
      <BannerMain className="aje-banner__main">{children}</BannerMain>
      {afterContent}
    </StyledBanner>
  );
}
