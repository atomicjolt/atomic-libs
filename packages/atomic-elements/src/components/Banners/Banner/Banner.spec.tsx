import { render, screen } from "@testing-library/react";
import { expect, it, describe } from "vitest";
import { Banner } from "./Banner.component";

describe("Banner", () => {
  describe("Snapshots", () => {
    it("should match snapshot", () => {
      const res = render(<Banner />);
      expect(res).toMatchSnapshot();
    });

    it("should match snapshot with variant", () => {
      const res = render(<Banner variant="success" />);
      expect(res).toMatchSnapshot();
    });
  });

  it("should forward props to the element", () => {
    render(<Banner className="test" data-attribute data-testid="aje-banner" />);
    expect(screen.getByTestId("aje-banner")).toBeInTheDocument();
    expect(screen.getByTestId("aje-banner")).toHaveClass("test");
    expect(screen.getByTestId("aje-banner")).toHaveAttribute("data-attribute");
  });

  describe("Banner.Content", () => {
    it("should match snapshot", () => {
      const res = render(<Banner.Content>Content</Banner.Content>);
      expect(res).toMatchSnapshot();
    });
  });

  describe("Banner.IconButton", () => {
    it("should match snapshot", () => {
      const res = render(<Banner.IconButton icon="more_vert" />);
      expect(res).toMatchSnapshot();
    });
  });

  describe("Banner.Button", () => {
    it("should match snapshot", () => {
      const res = render(<Banner.Button>Button</Banner.Button>);
      expect(res).toMatchSnapshot();
    });
  });
});
