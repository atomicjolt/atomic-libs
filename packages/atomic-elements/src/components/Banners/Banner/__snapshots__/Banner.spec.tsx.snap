// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Banner > Banner.Button > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="sc-deSjom byalIv aje-btn aje-btn--inverted is-auto"
        tabindex="0"
        type="button"
      >
        Button
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="sc-deSjom byalIv aje-btn aje-btn--inverted is-auto"
      tabindex="0"
      type="button"
    >
      Button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Banner > Banner.Content > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gjcoXW OnqIU"
      >
        Content
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gjcoXW OnqIU"
    >
      Content
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Banner > Banner.IconButton > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="sc-deSjom byalIv aje-btn sc-jHswkR WgpIQ aje-icon-btn aje-btn--banner is-medium"
        data-icon-button="true"
        tabindex="0"
        type="button"
      >
        <i
          aria-hidden="true"
          class="sc-fLDLck hYHdiZ aje-icon material-icons is-medium"
        >
          more_vert
        </i>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="sc-deSjom byalIv aje-btn sc-jHswkR WgpIQ aje-icon-btn aje-btn--banner is-medium"
      data-icon-button="true"
      tabindex="0"
      type="button"
    >
      <i
        aria-hidden="true"
        class="sc-fLDLck hYHdiZ aje-icon material-icons is-medium"
      >
        more_vert
      </i>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Banner > Snapshots > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-FEMpB eQljuX aje-banner aje-banner--info"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-FEMpB eQljuX aje-banner aje-banner--info"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Banner > Snapshots > should match snapshot with variant 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-FEMpB eQljuX aje-banner aje-banner--success"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-FEMpB eQljuX aje-banner aje-banner--success"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
